popgen/AFSsimulate.go

```diff
@@ -2,14 +2,89 @@ package popgen
 
 import (
 	"github.com/vertgenlab/gonomics/numbers"
+	"github.com/vertgenlab/gonomics/vcf"
+	"math/rand"
+	//DEBUG: "fmt"
 )
+
 //SimulateAFS returns an allele frequency spectrum AFS struct for n individuals with k segregating sites
 //with a selection parameter alpha.
-/*func SimulateAFS(alpha float64, n int, k int) AFS {
+func SimulateAFS(alpha float64, n int, k int) AFS {
 	var answer AFS
 	answer.sites = make([]*SegSite, 0)
+	var alleleFrequencies []float64 = StationaritySampler(alpha, k, 100000, 1000, 0.001, 0.999, true)
+	var count int
+	var r float64
 	for i := 0; i < k; i++ {
-		answer.sites = append(answer.sites, &SegSite{StationaritySample(n, alpha), n})
+		//now we simulate the discrete number of alleles for n individuals based on the allele frequency
+		count = 0
+		for j := 0; j < n; j++ {
+			r = rand.Float64()
+			if r < alleleFrequencies[i] {
+				count++ 
+			}
+		}
+		answer.sites = append(answer.sites, &SegSite{count, n})
+	}
+	return answer
+}
+
+//SimulateGenotype creates a matrix of GenomeSamples with k rows and n/2 columns, corresponding to k segregating sites and n alleles.
+//based on a selection parameter alpha. Alleles are shuffled among the individuals.
+func SimulateGenotype(alpha float64, n int, k int) [][]vcf.GenomeSample {
+	a := SimulateAFS(alpha, n, k)
+	var answer [][]vcf.GenomeSample = make([][]vcf.GenomeSample, 0, k)
+	
+	//now we initialize all column entries for each row of the answer matrix
+	for i := 0; i < len(answer); i++ {
+		answer[i] = make([]vcf.GenomeSample, 0, len(a.sites)/2 + 1)
+	}
+
+	var alleleArray []int16
+	var d int
+	var currRow []vcf.GenomeSample = make([]vcf.GenomeSample, 0, k)
```

Comment:
I think it makes sense to have the variable declared here, but I don't think you want to allocate the memory (make() after the equals sign) because you are just going to allocate different memory once you enter the loop in a few more lines.